### ISSUE TITLE
axe.run() overrides produce misleading Typescript warnings

### ISSUE THREAD

**Issue by ge-vlad-romanov** [dequelabs/axe-core#2316]
We are using the latest version of `axe-core` in our TS Protractor tests directly, since using `axe-webdriverjs`'s AxeBuilder throws TS compilation issues. To use `axe`, we declare it as a const of type `any`:

```
// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare const axe: any;
```
any we use it in a function like so:

```js
return browser.executeAsyncScript((opts: A11YOptions, done: (arg0: Promise<AxeResults>) => void) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
  const req: Promise<AxeResults> = axe.run(document, opts);
  done(req);
}, options);
```

Above, we provide it with the HTML document as the context and an options object with some rules to be disabled. Despite the code running, the TS compiler in the IDE warns that `axe.run()` expects 0 or 1 arguments, whereas in reality according to the overrides and the documentation it can accept anywhere between 0 and 3 (context, options and a callback).

Looking at the overloads for .run():
![Screen Shot 2020-06-19 at 10 34 19](https://user-images.githubusercontent.com/32774223/85145223-2a92b780-b244-11ea-81c0-5497975d5db2.png)

... we were wondering if it had something to do with the overloads ordering (https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#function-overloads), and if the function types should be more open by making use of optional params and union types?

**Comment by straker** [comment]
Thanks for the issue. We'll update the the types file before the 4.0 release and make sure it's correct and up-to-date. https://github.com/dequelabs/axe-core/issues/2317

**Comment by ge-vlad-romanov** [comment]
@straker Thanks for the quick response, I'm glad it was useful!

**Comment by straker** [comment]
Yep. It's planned for early August.

**Comment by straker** [comment]
@ge-vlad-romanov so I just [create a pr for the typescript file](https://github.com/dequelabs/axe-core/pull/2425) and made sure to test every possible combination of the `axe.run` call, and I couldn't reproduce this error.

Are you able to provide a live example I could look at?

**Comment by ge-vlad-romanov** [comment]
@straker Just to clarify my original post, the warning related to `axe.run()` is shown by the IDE (Webstorm), and it is not an actual TS compilation error, which may explain why you aren't able to reproduce it.

```ts
import {AxeResults, source} from 'axe-core';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare const axe: any;

public async runAccessibilityTest(): Promise<AxeResults> {
    interface A11YOptions {
      runOnly: {
        type: string;
        values: Array<string>;
      };
    }

    const options: A11YOptions = {
      runOnly: {
        type: 'tag',
        values: ['wcag2aa', 'wcag2a'],
      },
    };

    await browser.executeScript(source);
    return browser.driver.executeAsyncScript((opts: A11YOptions, done: (arg0: Promise<AxeResults>) => void) => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
      const req: Promise<AxeResults> = axe.run(document, opts);
      done(req);
    }, options);
  }
```

And this is the warning I get (I just reinstalled everything via npm ci just to make sure it still occurs):
![image](https://user-images.githubusercontent.com/32774223/88626229-34f56c00-d0a2-11ea-9d7e-40590741c568.png)

If it's actually something that is outside of your control then I am happy to live with it, but just wanted to double check with you guys.

**Comment by michael-siek** [comment]
Hey @ge-vlad-romanov if you change your import statement from 
```js
import {AxeResults, source} from 'axe-core';
```
to 
```js
import {run, AxeResults, source} from 'axe-core';
```

remove 
```js
// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare const axe: any;
```

and instead of using `axe.run()` use `run()`
```js
const req: Promise<AxeResults> =run(document, opts);
```

that should fix the issue above.

**Comment by ge-vlad-romanov** [comment]
Hi @michael-siek - thanks for taking a look. I actually did try that before without success, but I have now noticed that was because my declared type of `opts` was incompatible with what the overloads were expecting (I didn't see that axe provided a `RunOptions` interface, so I created my own, which is incompatible). I now have no warnings or errors :)

**Comment by straker** [comment]
Going to close as it looks like this is resolved. Please re-open if that is not the case.

**Comment by ge-vlad-romanov** [comment]
@straker @michael-siek Not sure if this is a problem on our side or with Protractor's interaction with axe-core, but using the above approach (importing `run` and using it directly) leads to this error when running the above script I posted:
```
JavascriptError: javascript error: axe_core_1 is not defined
JavaScript stack:
ReferenceError: axe_core_1 is not defined
    at eval (eval at executeAsyncScript (:556:26), <anonymous>:5:43)
    at eval (eval at executeAsyncScript (:556:26), <anonymous>:7:32)
    at eval (eval at executeAsyncScript (:556:26), <anonymous>:8:4)
    at executeAsyncScript (<anonymous>:556:47)
    at apply.runOnly.type (<anonymous>:571:29)
```
Thinking about it, this might have been why I had to initially go down the `declare const axe: any` approach and use `axe.run()`...

**Comment by javiergarciacotado** [comment]
Hello,

I have the same issue of this latest post ... How did you solve it?
Thanks,

**Comment by sumerjoshi** [comment]
@ge-vlad-romanov How did you end up solving this? @javiergarciacotado

**Comment by borisdiakur** [comment]
For those who come here after searching for `axe_core_1 is not defined`, here's what I did to resolve this in my project:

Instead of `import axe from 'axe-core'` use `import * as axe from 'axe-core'`
